### PR TITLE
Multiple density estimate plot

### PR DESCRIPTION
**Pull request by Azaya89**
- [x] issues: addresses #13364 as a continuous issue
- [x] tests added / passed

### CODE REVIEW

File: docs/bokeh/source/docs/user_guide/topics/stats.rst [hunk 1]
```diff
@@ -34,6 +34,13 @@ Kernel density estimation
 
 .. _ug_topics_stats_sinaplot:
 
+Kernel density estimates can also be created using |varea| glyphs:
```
Comment (bryevdv) [inline docs/bokeh/source/docs/user_guide/topics/stats.rst]:
```suggestion
Kernel density estimates can also be plotted using |varea| glyphs:
```
(scikit-learn is actually what is *creating* the estimates)

File: examples/topics/stats/density.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+from bokeh.models import CustomJSTickFormatter, Label
+from bokeh.plotting import figure, show
+from bokeh.sampledata.cows import data as df
+
+grouped = df.groupby('breed')
+
+butterfat = []
+for breed, group_df in grouped:
+    values = group_df['butterfat'].values
+    butterfat.append(values)
```
Comment (bryevdv) [inline examples/topics/stats/density.py]:
```suggestion
    butterfat.append(group_df['butterfat'].values)
```

File: examples/topics/stats/density.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+new_dict = {
+    "values": [butterfat[0], butterfat[1], butterfat[2],butterfat[3], butterfat[4]],
```
Comment (bryevdv) [inline examples/topics/stats/density.py]:
```suggestion
    "values": butterfat,
```
?

File: examples/topics/stats/density.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    "bandwidths": [0.2, 0.2, 0.2, 0.2, 0.2],
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: If we are going to use the same value for every case then there is no need for this column, it can be removed here and code simplified below&mdash;just pass `bandwidth=0.2` to `KernelDensity` instead

File: examples/topics/stats/density.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    "colors": ["#409DFA", "#FF3399", "#AC5703", "#9E5205", "green"],
+    "labels": ["Ayrshire", "Canadian", "Guernsey", "Holstein-Friesian", "Jersey"],
+}
+
+new_df = pd.DataFrame(new_dict)
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: the `new_`'s here seem superfluous (there aren't any "old" values?) Suggest reworking to look like:
```
df = p.DataFrame(data)
```

File: examples/topics/stats/density.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+p = figure(title="Density estimates of butterfat percentages in cows", height=400, x_axis_label="butterfat contents", y_axis_label="density")
+
+positions = np.linspace(2, 8, 1000)
+for _, row in new_df.iterrows():
+    data, bandwidth, color, label = (row["values"], row["bandwidths"], row["colors"], row["labels"])
+
+    kde = KernelDensity(kernel="gaussian", bandwidth=bandwidth).fit(data[:, np.newaxis])
+    log_dens = kde.score_samples(positions[:, np.newaxis])
+
+    p.varea(x=positions, y1=np.exp(log_dens), y2=0, fill_alpha=0.3, fill_color=color)
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: This sends a duplicate copy of `positions` for every `varea`.  Since the series length is the same for every case, it would be preferable to construct one CDS for all the calls to `varea`:
```python
p.varea(x="x", y1=label, y2=0, fill_alpha=0.3, fill_color=color, source=source)
```
where `source.data[label]` has the y-data for that label
```

File: examples/topics/stats/density.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        text_font_size="10pt",
+        x_offset=10,
+        y_offset=-5,
+        text_color=color,
+    )
+    p.add_layout(highest_point_label)
+
+# Convert x-axis labels to percentages
+x_axis_labels = {3: "3%", 4: "4%", 5: "5%", 6: "6%", 7: "7%"}
+p.xaxis.formatter = CustomJSTickFormatter(code="""var labels = %s; return labels[tick] || '';""" % x_axis_labels)
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: I don't think we need a `CustomJSTickFormatter` or `x_axis_labels` here? I think you can just use built-in `NumeralTickFormatter` similar to the example in the docs:

https://docs.bokeh.org/en/latest/docs/user_guide/styling/plots.html#numeraltickformatter

File: examples/topics/stats/density.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+p.axis.axis_line_color = None
+p.axis.major_tick_out = 0
+p.axis.minor_tick_out = 0
+p.axis.major_tick_in = 0
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: you should be able to set `p.axis.major_tick_line_color` and `p.axis.minor_tick_line_color` both to None and then not worry about "tick_in" etc which are a bit obscure. IMO setting the colors to None is a more clear way to indicate "no ticks are desired"

File: examples/topics/stats/density.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+p.x_range.start = 2.5
+p.x_range.end = 7.5
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: suggest passing `x_range = (2.5, 7.5)` directly to `figure` instead of setting after the fact.

File: examples/topics/stats/density.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+p.xgrid.grid_line_color = None
+p.yaxis.ticker = [0, 0.5, 1, 1.5]
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: Does a tuple work here? If not, please file an issue. If a tuple does work please change to a tuple (always good to default prefer immutable data structures whenever feasible)

**Review by bryevdv — COMMENTED** [review]
Some further comments

File: examples/topics/stats/density.py (hunk 7 above)
Comment (Azaya89) [reply examples/topics/stats/density.py]: The `NumeralTickFormatter` _converts_ 1 to percentage (from 1 to "100%") when used like this:
`p.xaxis.formatter = NumeralTickFormatter(format="0%")` but what I want is simply to _rewrite_ the numbers as percentages, from 1 to "1%". I'm not sure how to do that with NTF.

Comment (bryevdv) [reply examples/topics/stats/density.py]: In that case I think the [`PrintfTickFormatter`](https://docs.bokeh.org/en/latest/docs/user_guide/styling/plots.html#printftickformatter) would also be an option, with a format like `"%d%%"` (where `%d` is the format for ints and `%%` is the escape for a literal `%` character)

**Comment by bryevdv** [comment]
@Azaya89 I see all the conversations marked resolved but no new commits, did you intend to push changes for those comments?

**Comment by Azaya89** [comment]
> @Azaya89 I see all the conversations marked resolved but no new commits, did you intend to push changes for those comments?

Yes, but I wanted to resolve the one about the tick formatting before pushing all of them together.

File: examples/topics/stats/density.py (hunk 7 above)
Comment (Azaya89) [reply examples/topics/stats/density.py]: Thank you.

File: examples/topics/stats/density.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+colors = Dark2[5]
+x = np.linspace(2, 8, 1000)
+source = ColumnDataSource(dict(x=x))
+
+p = figure(title="Multiple density estimates", height=300, x_range=(2.5, 7.5), x_axis_label="butterfat contents", y_axis_label="density")
+
+for i, (breed, group_df) in enumerate(grouped):
+    data = group_df['butterfat'].values
+    kde = KernelDensity(kernel="gaussian", bandwidth=0.2).fit(data[:, np.newaxis])
+    log_dens = kde.score_samples(x[:, np.newaxis])
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: we have enough space to avoid abbreviations
```suggestion
    log_density = kde.score_samples(x[:, np.newaxis])
```

File: examples/topics/stats/density.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+from bokeh.sampledata.cows import data as df
+
+grouped = df.groupby('breed')
+colors = Dark2[5]
+x = np.linspace(2, 8, 1000)
+source = ColumnDataSource(dict(x=x))
+
+p = figure(title="Multiple density estimates", height=300, x_range=(2.5, 7.5), x_axis_label="butterfat contents", y_axis_label="density")
+
+for i, (breed, group_df) in enumerate(grouped):
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: Just for consideration, I think the enumeration and explicit loop index could be avoided with:
```python
for (breed, breef_df), color in zip(grouped, colors):
```
in any case I think think `breed_df` is a a bit more descriptive than `group_df` (and maybe `grouped` could be something like `by_breed` or `breed_groups`?)

File: examples/topics/stats/density.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        y=y[max_idx],
+        text=breed,
+        text_font_size="10pt",
+        x_offset=10,
+        y_offset=-5,
+        text_color=colors[i],
+    )
+    p.add_layout(highest_point_label)
+
+# Convert x-axis labels to percentages
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: slightly more accurate:
```suggestion
# Display x-axis labels as percentages
```

File: examples/topics/stats/density.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+for i, (breed, group_df) in enumerate(grouped):
+    data = group_df['butterfat'].values
+    kde = KernelDensity(kernel="gaussian", bandwidth=0.2).fit(data[:, np.newaxis])
+    log_dens = kde.score_samples(x[:, np.newaxis])
+    y = np.exp(log_dens)
+    source.add(y, breed)
+    p.varea(x="x", y1=breed, y2=0, source=source, fill_alpha=0.3, fill_color=colors[i])
+
+    # Find the highest point and annotate with the label
```
Comment (bryevdv) [inline examples/topics/stats/density.py]:
```suggestion
    # Find the highest point and annotate with a label
```

File: examples/topics/stats/density.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+import numpy as np
+from sklearn.neighbors import KernelDensity
+
+from bokeh.models import ColumnDataSource, Label, PrintfTickFormatter
+from bokeh.palettes import Dark2
+from bokeh.plotting import figure, show
+from bokeh.sampledata.cows import data as df
+
+grouped = df.groupby('breed')
+colors = Dark2[5]
```
Comment (bryevdv) [inline examples/topics/stats/density.py]: I think you can get rid of this line this by importing this way at the top:
```
from bokeh.palettes import Dark2_5 as colors
```

**Review by bryevdv — COMMENTED** [review]
Looking really good now, just a couple of final observations

**Review by bryevdv — APPROVED** [review]
Mypy failures are unrelated. Great example!
